interfaces/many: conditionally use 'unsafe' with change_profile rules

jdstrand opened the pull request:
* interfaces/docker-support: add unsafe change_profile rules
* conditionally apply 'ix' rule to home interface to avoid LP: #1797786
 Certain change_profile rules comflict with the 'ix' rules in the home interface. By default, always add 'ix' but allow interfaces to request suppressing the use of 'ix'

References:
* https://bugs.launchpad.net/apparmor/+bug/1797786
* https://forum.snapcraft.io/t/ld-library-path-in-snapped-docker/6903/16
* https://github.com/kubeflow/kubeflow/issues/1367

Note: this is branched from https://github.com/snapcore/snapd/pull/5981 which was branched from https://github.com/snapcore/snapd/pull/5980

zyga commented on interfaces/apparmor/spec.go:
```diff
@@ -57,6 +57,14 @@ type Specification struct {
+	// require this type of change_profile rule to suppress 'ix' so that
+	// the calling interface can be used with the home interface. Ideally,
+	// we would not need this, but we currently do (LP: #1797786)
+	suppressHomeIx bool
```
+1 on the change with the same suggestion to name this differently as in the PR that introduces ptrace methods and variables.

jdstrand replied:
Done

zyga commented on interfaces/apparmor/spec.go:
```diff
@@ -509,3 +517,12 @@ func (spec *Specification) GetUsesPtraceTrace() bool {
+	spec.suppressHomeIx = true
+}
+
+func (spec *Specification) GetSuppressHomeIx() bool {
```
As in the other PR I would suggest to discard `GetSupressesHomeIx` as this is only used by the back-end code with has access to the private attribute.

jdstrand replied:
I use the Getter in common_test.go and felt this was cleaner.

anonymouse64 commented:
Just wanted to comment that I tested this change with the most recent docker snap released (18.06) and setting `LD_LIBRARY_PATH` works now. Thanks!

```
$ docker run  -e "LD_LIBRARY_PATH=/foo" -e "MY_VAR=is_set" jupyterhub/jupyterhub:0.8.1 env
PATH=/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=51e373c8c575
MY_VAR=is_set
LD_LIBRARY_PATH=/foo
DEBIAN_FRONTEND=noninteractive
LANG=C.UTF-8
HOME=/root
```

jdstrand commented:
@zyga - I answered/addressed all feedback and think this is ready for another review.

mvo5 commented on interfaces/apparmor/backend_test.go:
```diff
@@ -1660,3 +1660,50 @@ func (s *backendSuite) TestPtraceTraceRule(c *C) {
+		}
+
+		snapInfo := s.InstallSnap(c, tc.opts, "", ifacetest.SambaYamlV1, 1)
+		s.parserCmd.ForgetCalls()
```
Is this needed? It seems the test never checks for s.parserCmd

mvo5 commented on interfaces/apparmor/backend_test.go:
```diff
@@ -1660,3 +1660,50 @@ func (s *backendSuite) TestPtraceTraceRule(c *C) {
+		snapInfo := s.InstallSnap(c, tc.opts, "", ifacetest.SambaYamlV1, 1)
+		s.parserCmd.ForgetCalls()
+
+		err := s.Backend.Setup(snapInfo, tc.opts, s.Repo)
```
AFACT this is already done in s.InstallSnap() so probably not needed again.

zyga replied:
I pushed a fix for this.

mvo5 commented on interfaces/apparmor/backend_test.go:
```diff
@@ -1660,3 +1660,50 @@ func (s *backendSuite) TestPtraceTraceRule(c *C) {
+		profile := filepath.Join(dirs.SnapAppArmorDir, "snap.samba.smbd")
+		data, err := ioutil.ReadFile(profile)
+		c.Assert(err, IsNil)
+
```
(nitpick^2) This whitespace feels a bit out of place, the two asserts look related.

zyga commented:
I had a feeling I reviewed this but obviously that is just because I did in one of the earlier branches.

I will review it again, trying to cross reference the old discussion.

zyga commented:
I restarted tests a number of times. Let's see if the failure is real or just a coincidence.

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -562,6 +562,14 @@ func addContent(securityTag string, snapInfo *snap.Info, opts interfaces.Confine
 				if spec.SuppressPtraceTrace() && !spec.UsesPtraceTrace() {
 					tagSnippets += ptraceTraceDenySnippet
 				}
+
```
This is fine

zyga commented on interfaces/apparmor/spec.go:
```diff
@@ -509,3 +517,12 @@ func (spec *Specification) SetSuppressPtraceTrace() {
 }
+
+// SetSuppressHomeIx to request explicit ptrace deny rules
+func (spec *Specification) SetSuppressHomeIx() {
```
This is fine

zyga commented on interfaces/builtin/common.go:
```diff
@@ -94,6 +95,9 @@ func (iface *commonInterface) AppArmorConnectedPlug(spec *apparmor.Specification
 	} else if iface.suppressPtraceTrace {
 		spec.SetSuppressPtraceTrace()
 	}
+	if iface.suppressHomeIx {
```
This is fine

zyga commented on interfaces/builtin/docker_support.go:
```diff
@@ -568,12 +570,37 @@ func (iface *dockerSupportInterface) StaticInfo() interfaces.StaticInfo {
 	var privileged bool
 	_ = plug.Attr("privileged-containers", &privileged)
-	spec.AddSnippet(dockerSupportConnectedPlugAppArmor)
+	useUnsafe := false
```
This can use a helper from `strutils`
```
useUnsafe := strutil.ListContains(parserFeatures(), "unsafe")
```

zyga commented on interfaces/builtin/docker_support.go:
```diff
@@ -568,12 +570,37 @@ func (iface *dockerSupportInterface) StaticInfo() interfaces.StaticInfo {
+			useUnsafe = true
+		}
+	}
+	rule := "change_profile -> docker-default,"
```
This is fine

zyga commented:
I will need to review and perhaps alter this branch with the lessons learned about openSUSE Leap 42.3 and apparmor parser there.

jdstrand commented:
@mvo5, @zyga (cc @pedronis) - I believe I addressed all of the feedback. Now that we require "unsafe" elsewhere (https://github.com/snapcore/snapd/blob/master/release/apparmor.go#L200), one could argue that I could lose all the conditional logic and just unconditionally use "unsafe" in the rules for the interface (this might be what @zyga was referring to in his previous comment). Happy to simplify if others feel I should.

zyga commented:
@jdstrand can't have the cake and eat it too. I'm +1 on simplify

jdstrand commented:
@mvo5 - this has two approvals but you had some things you wanted addressed (which I addressed), so don't know if you want to take another look before committing.
